### PR TITLE
Add HubConnection receive benchmark

### PR DESCRIPTION
**Pull request by JamesNK**
https://github.com/aspnet/SignalR/issues/2126

### CODE REVIEW

**Comment by JamesNK** [comment]
```
       Method | ArgumentCount | MessageCount |    Protocol |         Mean |      Error |     StdDev |      Op/s | Allocated |
------------- |-------------- |------------- |------------ |-------------:|-----------:|-----------:|----------:|----------:|
 ReceiveAsync |             0 |            1 |        json |     1.989 us |  0.0285 us |  0.0238 us | 502,657.1 |     920 B |
 ReceiveAsync |             0 |            1 | messagepack |     1.089 us |  0.0161 us |  0.0143 us | 917,979.5 |     432 B |
 ReceiveAsync |             0 |          100 |        json |   165.248 us |  1.8490 us |  1.7295 us |   6,051.5 |   69824 B |
 ReceiveAsync |             0 |          100 | messagepack |    80.157 us |  1.6074 us |  1.5036 us |  12,475.6 |   21024 B |
 ReceiveAsync |             1 |            1 |        json |     2.231 us |  0.0366 us |  0.0376 us | 448,311.5 |    1064 B |
 ReceiveAsync |             1 |            1 | messagepack |     1.229 us |  0.0125 us |  0.0098 us | 813,846.0 |     496 B |
 ReceiveAsync |             1 |          100 |        json |   189.748 us |  1.5653 us |  1.4642 us |   5,270.2 |   84224 B |
 ReceiveAsync |             1 |          100 | messagepack |    92.179 us |  1.7636 us |  1.8111 us |  10,848.5 |   27424 B |
 ReceiveAsync |            10 |            1 |        json |     4.450 us |  0.0867 us |  0.0963 us | 224,713.4 |    2144 B |
 ReceiveAsync |            10 |            1 | messagepack |     2.216 us |  0.0306 us |  0.0271 us | 451,323.9 |    1072 B |
 ReceiveAsync |            10 |          100 |        json |   425.907 us |  4.2827 us |  3.3436 us |   2,347.9 |  192224 B |
 ReceiveAsync |            10 |          100 | messagepack |   184.709 us |  0.5861 us |  0.4894 us |   5,413.9 |   85024 B |
 ReceiveAsync |           100 |            1 |        json |    26.164 us |  0.5031 us |  0.4706 us |  38,220.7 |   12944 B |
 ReceiveAsync |           100 |            1 | messagepack |    11.426 us |  0.2204 us |  0.3090 us |  87,520.8 |    6832 B |
 ReceiveAsync |           100 |          100 |        json | 2,534.927 us |  9.4322 us |  7.3641 us |     394.5 | 1272224 B |
 ReceiveAsync |           100 |          100 | messagepack | 1,101.649 us | 16.3335 us | 13.6392 us |     907.7 |  661024 B |
```

File: benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/HubConnectionReceiveBenchmark.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        public string Protocol;
+
+        [GlobalCleanup]
+        public void GlobalCleanup()
+        {
+            _nextReadTcs.SetResult(new ReadResult(default, false, true));
+            _hubConnection.StopAsync().GetAwaiter().GetResult();
+        }
+
+        public void OperationSetup()
```
Comment (BrennanConroy) [inline benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/HubConnectionReceiveBenchmark.cs]: Make this `[IterationSetup]` so it doesn't count as part of the actual benchmark run.

Comment (JamesNK) [reply benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/HubConnectionReceiveBenchmark.cs]: This originally was called for `IterationSetup` but each method call is an operation, and many operations make up an iteration, so it didn't work.

https://github.com/dotnet/BenchmarkDotNet/blob/master/docs/guide/Advanced/SetupAndCleanup.md#iterationsetup

Comment (BrennanConroy) [reply benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/HubConnectionReceiveBenchmark.cs]: Oh interesting :(

**Comment by JamesNK** [comment]
I tried reverting https://github.com/aspnet/SignalR/pull/2136/files and the benchmarks don't even run with it back again. I think something about it is causing threadpool starvation.

**Comment by JamesNK** [comment]
With `await Task.Delay(1)`
```
       Method | ArgumentCount | MessageCount | Protocol |     Mean |     Error |    StdDev |  Op/s | Allocated |
------------- |-------------- |------------- |--------- |---------:|----------:|----------:|------:|----------:|
 ReceiveAsync |             0 |            1 |     json | 15.58 ms | 0.0262 ms | 0.0233 ms | 64.17 |   2.27 KB |
```
Without
```
       Method | ArgumentCount | MessageCount | Protocol |     Mean |     Error |    StdDev |      Op/s |  Gen 0 | Allocated |
------------- |-------------- |------------- |--------- |---------:|----------:|----------:|----------:|-------:|----------:|
 ReceiveAsync |             0 |            1 |     json | 2.031 us | 0.0400 us | 0.0476 us | 492,303.8 | 0.0076 |     920 B |
```

File: benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/HubConnectionReceiveBenchmark.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        {
+            _tcs = _nextReadTcs;
+
+            // Add the results for additional messages (minus 1 because 1 result has already been added)
+            for (int i = 0; i < MessageCount - 1; i++)
+            {
+                _pipe.AddReadResult(new ValueTask<ReadResult>(new ReadResult(new ReadOnlySequence<byte>(_invocationMessageBytes), false, false)));
+            }
+
+            // The reveice task that will be waited on once messages are read
```
Comment (BrennanConroy) [inline benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/HubConnectionReceiveBenchmark.cs]: receive

**Comment by JamesNK** [comment]
@dotnet-bot test Windows Release x64 Build

**Pull request merged**